### PR TITLE
Implement Bulk elasticsearch

### PR DESCRIPTION
**Pull request by 1Jo1**
# Description

elastic-indexer is responsible to collect all messages for a specific operation and to send it to elasticsearch 

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] posts-indexer tested on kubernetes

### CODE REVIEW

File: elastic/indexer/elastic-indexer.go [hunk 1]
```diff
@@ -23,24 +24,26 @@ import (
...
+func New(esHosts []string, esIndex, esMapping, kafkaAddress, changesTopic, kafkaGroupID string, indexFunc IndexFunc, bulkSize int) *Indexer {
 	readerConfig := kf.NewReaderConfig(kafkaAddress, kafkaGroupID, changesTopic)
 
 	i := &Indexer{}
 	i.kReader = kf.NewReader(readerConfig)
 	i.indexFunc = indexFunc
-
+	i.esIndex = esIndex
 	i.esClient = elastic.InitializeElasticSearch(esHosts)
+	i.bulkSize = bulkSize
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: Let's call this `bulkChunkSize`

File: elastic/indexer/elastic-indexer.go [hunk 2]
```diff
@@ -52,20 +55,84 @@ func New(esHosts []string, esIndex, esMapping, kafkaAddress, changesTopic, kafka
...
 		return err
 	}
-	return i.kReader.CommitMessages(context.Background(), m)
+
+	log.Println("BulkResultItem: ", len(result.Items))
+	log.Println("BulkResult: ", string(body))
+	for _, bulkResultOperation := range result.Items {
+
+		if bulkResultOperation.Index.ID != "" {
+			err := i.commitKafkaMessages(bulkResultOperation.Index.Status, bulkResultOperation.Index.ID, bulkDocumentIdKafkaMessages)
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: As said in person: 
Be careful with committing when one of the items contains an error

File: elastic/indexer/elastic-indexer.go [hunk 3]
```diff
@@ -121,6 +127,23 @@ func (i *Indexer) runStep() error {
 	return nil
 }
 
+func (i *Indexer) checkAllResultMessagesAreValid(result *bulkResult) error {
+	for _, bulkResultOperation := range result.Items {
+		if bulkResultOperation.Index.ID != "" {
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: If you define the type of `Index` as `*struct{...}` you can check if the pointer is nil ;) 
`json.Unmarshal` will still work correctly

File: elastic/indexer/elastic-indexer.go [hunk 4]
```diff
@@ -121,6 +127,23 @@ func (i *Indexer) runStep() error {
...
+			return checkHttpStatus(bulkResultOperation.Index.Status)
+		} else if bulkResultOperation.Update.ID != "" {
+			return checkHttpStatus(bulkResultOperation.Update.Status)
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: This way`checkAllResultMessagesAreValid` will return `error`/`nil` from the first item, you should only return if the `error` returned by  `checkHttpStatus ` is not nil

File: elastic/indexer/elastic-indexer.go [hunk 5]
```diff
@@ -121,6 +127,23 @@ func (i *Indexer) runStep() error {
...
+		}
+	}
+}
+
+func checkHttpStatus(httpStatus int) error {
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: I would expect `checkHttpStatus` to return a bool
What do you think about calling this `ErrorForStatus`?
That would make the calling code quite expressive

**Review by alexmorten — CHANGES_REQUESTED** [review]

File: elastic/indexer/elastic-indexer.go (hunk 5 above)
Comment (1Jo1) [reply elastic/indexer/elastic-indexer.go]: `ErrorForHttpStatus` sounds good

File: elastic/indexer/elastic-indexer.go [hunk 6]
```diff
@@ -52,20 +55,117 @@ func New(esHosts []string, esIndex, esMapping, kafkaAddress, changesTopic, kafka
...
+
+func errorForHttpStatus(httpStatus int) error {
+	if httpStatus != 200 && httpStatus != 201 {
+		return fmt.Errorf("creating/updateing index failed Httpstatus: %d \n", httpStatus)
+	}
+	return nil
+}
+
+func (i *Indexer) commitKafkaMessages(httpStatus int, documentId string, bulkDocumentIdKafkaMessages map[string]kafka.Message) error {
+	if httpStatus == 200 || httpStatus == 201 {
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: We don't need to check the httpStatus anymore, right? 
When  `commitKafkaMessages` is called, the items have already been checked 
-> I don't think we need a helper func for commiting the messages anymore :)

Comment (1Jo1) [reply elastic/indexer/elastic-indexer.go]: haha that's true  :)

File: elastic/indexer/elastic-indexer.go [hunk 7]
```diff
@@ -52,20 +55,105 @@ func New(esHosts []string, esIndex, esMapping, kafkaAddress, changesTopic, kafka
...
+
+	err = i.checkAllResultMessagesAreValid(&result)
+	if err != nil {
 		return err
 	}
-	return i.kReader.CommitMessages(context.Background(), m)
+
+	for _, bulkResultOperation := range result.Items {
+
+		if bulkResultOperation.Index.ID != "" {
```
Comment (alexmorten) [inline elastic/indexer/elastic-indexer.go]: you should also change this to `bulkResultOperation.Index != nil` :)

File: elastic/indexer/model.go [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package indexer
+
+type ElasticIndexer struct {
```
Comment (alexmorten) [inline elastic/indexer/model.go]: I don't think `ElasticIndexer` describes the purpose of this struct well; what do you think about something like `BulkIndexDoc`?

**Review by alexmorten — APPROVED** [review]
Looks good! 
Just one tiny thing, feel free to merge afterwards

File: elastic/indexer/model.go (hunk 8 above)
Comment (1Jo1) [reply elastic/indexer/model.go]: is much better :)   I couldn't think of anything better :)

**Pull request merged**